### PR TITLE
Port point cloud transformation to numpy

### PR DESCRIPTION
**Pull request by Flova**
Implements the point cloud transformation in NumPy. This speeds up the transformation operation by 20000% for large point clouds (excluding the serialization and deserialization). The code needs some further testing but should work as it is. I also try to improve the performance of the serialization and deserialization as seen in https://github.com/ros2/common_interfaces/pull/175.

This adds NumPy as a dependency and drops the dependency on PyKDL which should solve #471.

The API is also improved by adding proper interface descriptions.

Cheers

### CODE REVIEW

**Comment by aprotyas** [comment]
CI is failing here because it couldn't resolve the `python-numpy` rosdep key for Ubuntu Jammy. I believe the dependency needs to be on `python3-numpy` instead.

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: Can you rearrange the imports in alphabetic order?

File: tf2_sensor_msgs/package.xml
Comment (aprotyas) [inline tf2_sensor_msgs/package.xml]:
```suggestion
  <depend>python3-numpy</depend>
```

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: I would suggest considering the PEP-484-style type annotations for numpy ([`numpy.typing`](https://numpy.org/devdocs/reference/typing.html)).

```python3
import numpy.typing as npt

def transform_points(point_count: npt.ArrayLike, transform: Transform) -> npt.ArrayLike:
    ...
```

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: Same type annotation suggestion here, and also annotate the return type.

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: Math LGTM, might want to leave a pointer to the reference implementation?

https://github.com/matthew-brett/transforms3d/blob/f185e866ecccb66c545559bc9f2e19cb5025e0ab/transforms3d/quaternions.py#L101

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: Single quotes preferred over double quotes, from the style guide: https://docs.ros.org/en/rolling/Contributing/Code-Style-Language-Versions.html#id4

```suggestion
    return np.einsum('ij, pj -> pi', transform_rotation_matrix, point_cloud + transform_translation)
```

**Review by aprotyas — CHANGES_REQUESTED** [review]
This looks like a great contribution. I don't know about adding numpy as a dependency, specially since upstream efforts for PyKDL are underway, but the maintainers can decide on that. I've left some suggestions inline.

Also, it looks like your commits aren't signed-off. Do you mind signing them off?
https://docs.ros.org/en/rolling/Contributing/Developer-Guide.html#change-control-process

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [reply tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: i.e.
```suggestion
def _get_mat_from_quat(quaternion: npt.ArrayLike) -> npt.ArrayLike:
```

**Comment by aprotyas** [comment]
Also, is the DCO bot not configured on this repo for some reason? @clalancette

**Comment by Flova** [comment]
> CI is failing here because it couldn't resolve the `python-numpy` rosdep key for Ubuntu Jammy. I believe the dependency needs to be on `python3-numpy` instead.

Done

**Comment by Flova** [comment]
There seems to be an issue with `/usr/lib/x86_64-linux-gnu/libpython3.9.so` on jammy. 
I get the following error while building:

```
No rule to make target '/usr/lib/x86_64-linux-gnu/libpython3.9.so', needed by 'libstd_msgs__rosidl_generator_c.so'.  Stop.
```

And the CI fails with a similar one.

**Comment by Flova** [comment]
Sorry I needed to force push the new commit with signature.

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (timonegk) [reply tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: If I see it correctly, the calculation of Nq is wrong, here it is the norm which is the square root of the sum of squares, it should be the sum of squares instead.

Comment (Flova) [reply tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: Oh good catch

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (Flova) [reply tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: `npt.ArrayLike` is defined as everything that can be cast into a NumPy array using e.g. the `np.asarray()` method. This included scalars as well as lists, NumPy arrays, and so on. So we could extend this interface to include all of these, but we would also need to add an `np.asarray()` call for the input. I think it is okay to type it as a `np.ndarray`, because it is a private function which is only used with NumPy arrays. 

Also the output is strictly a `np.ndarray` and not a `npt.ArrayLike`

See: https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (Flova) [reply tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: The function returns an `np.ndarray` which is more specific than an `npt.ArrayLike`. I added the `npt.ArrayLike` type to the input point cloud, but this requires a typecast from any array like object to a `np.ndarray` for further processing. This is done via `np.asarray` and should come with nearly no performance implications in the input is already a `np.ndarray`.

**Comment by Flova** [comment]
All tests are passing. But I would want to wait for the serialization pull request, which would open some possibilities for improvements in this pr.

Any further suggestions regarding this implementation?

File: tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (Flova) [reply tf2_sensor_msgs/src/tf2_sensor_msgs/tf2_sensor_msgs.py]: [numpy.typing](https://numpy.org/devdocs/reference/typing.html) is sadly not available for Ubuntu 20.04 we therefore need to find another solution. Discussed in PR #509.

File: tf2_sensor_msgs/CMakeLists.txt
Comment (aprotyas) [inline tf2_sensor_msgs/CMakeLists.txt]: This line isn't needed because this is what `PACKAGE_DIR` would have defaulted to.

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: If the public API is imported into the `tf2_sensor_msgs/__init__.py` scope, you can do this:
```suggestion
from tf2_sensor_msgs import do_transform_cloud, transform_points
```

File: tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py]: Do you plan to work on these TODOs for this PR? If not, I suggest removing these comments.

**Review by aprotyas — COMMENTED** [review]
The PR looks good mostly. I've left some small comments in-line.

It might make sense to import the API in the scope of `tf2_sensor_msgs/__init__.py` with:

```python3
from .tf2_sensor_msgs import do_transform_cloud, transform_points  # noqa
```

**Comment by aprotyas** [comment]
@ros-pull-request-builder retest this please

File: tf2_sensor_msgs/package.xml
Comment (aprotyas) [inline tf2_sensor_msgs/package.xml]:
```suggestion
  <depend>geometry_msgs</depend>
  <depend>python3-numpy</depend>
  <depend>sensor_msgs</depend>
  <depend>sensor_msgs_py</depend>
  <depend>tf2</depend>
  <depend>tf2_ros</depend>
```

**Comment by aprotyas** [comment]
CI:
Build args: `--packages-above-and-dependencies tf2_sensor_msgs`
Test args: `--packages-above tf2_sensor_msgs`
Repos file: https://gist.githubusercontent.com/aprotyas/a62e4da45b1ec3f5d52838ff94617e94/raw/a61329ea9bfded0ef42a4d5ccae47b321d6dd937/ros2.repos
Job: https://ci.ros2.org/job/ci_launcher/10079

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16463/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11053/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16815/)

**Comment by aprotyas** [comment]
@Flova looks like there are some linter errors in CI: can you address that?
Also, one of the tests fail because `sensor_msgs_py` can't be imported. I believe updating the list of dependencies (https://github.com/ros2/geometry2/pull/507#discussion_r837200059) should sort that out.

**Comment by clalancette** [comment]
Also, it looks like some of the numpy changes from https://github.com/ros2/common_interfaces/pull/175 caused regressions on RHEL: https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_release/1091/ .  I would suggest holding off on this PR until we figure out what is going on there and how to mitigate it (as we may need to do the same thing here).

**Comment by aprotyas** [comment]
> Also, it looks like some of the numpy changes from [ros2/common_interfaces#175](https://github.com/ros2/common_interfaces/pull/175) caused regressions on RHEL: https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_release/1091/ . I would suggest holding off on this PR until we figure out what is going on there and how to mitigate it (as we may need to do the same thing here).

Related to https://github.com/ros2/common_interfaces/pull/185, the `numpy.typing` module is introduced in numpy 1.20, and so will also cause regressions on RHEL.

**Comment by Flova** [comment]
> > Also, it looks like some of the numpy changes from [ros2/common_interfaces#175](https://github.com/ros2/common_interfaces/pull/175) caused regressions on RHEL: https://ci.ros2.org/view/nightly/job/nightly_linux-rhel_release/1091/ . I would suggest holding off on this PR until we figure out what is going on there and how to mitigate it (as we may need to do the same thing here).
> 
> Related to [ros2/common_interfaces#185](https://github.com/ros2/common_interfaces/pull/185), the `numpy.typing` module is introduced in numpy 1.20, and so will also cause regressions on RHEL.

Not only for RHEL. Afaik Ubuntu has a recent version that doesn't support it too. So i will remove the npt typing.

File: tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (Flova) [reply tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py]: I plan to resolve them for this pr. But they needed the changes in https://github.com/ros2/common_interfaces/pull/175, and I was busy since it was merged.

Comment (Flova) [reply tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py]: I resolved the last TODOs

**Comment by Flova** [comment]
All TODOs should be resolved now. I added 

- the RHEL fix
- optimizations if the cloud only includes coordinates
- more tests for clouds that contain not only coordinates
- more tests regarding the expected header as well as the registration in tf2

The rolling CI also passes, so we should be fine to rerun the tests for all architectures. @aprotyas 
Maybe also have a look at my additions to the as it changed a bit.

File: tf2_sensor_msgs/tf2_sensor_msgs/__init__.py
Comment (aprotyas) [inline tf2_sensor_msgs/tf2_sensor_msgs/__init__.py]: Do you want to be importing everything into this scope? I think `transform_points` and `do_transform_cloud` are the only public API right?

Maybe consider something like:
```suggestion
from .tf2_sensor_msgs import do_transform_cloud
from .tf2_sensor_msgs import transform_points

__all__ = [
    'do_transform_cloud',
    'transform_points',
]
```

**Comment by Flova** [comment]
Friendly ping @aprotyas :)

File: tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py:37]: I don't have any suggestions here, just pointing something out:

I don't think `numpy_compat.[un]structured_to_[un]structured` should be part of `sensor_msgs_py`'s public API, but that shouldn't be part of this discussion I guess. It just feels off to use utility functions that should be housed somewhere else, but I think it can be rationalized by the fact that there isn't critical mass/use to put this numpy utility elsewhere.

File: tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py:87]: Neat!

File: tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py]: Nit: I believe this works too, and we don't incur list construction costs
```suggestion
    required_fields = set('xyz')
```

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Let's not put in this assertion.

By construction, functional code and test code are tested one against the other, and if the point cloud does not contain "our points", the tests won't pass anyway.

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: This can just be set to zero right? Same remark for line 104 and 123.

```suggestion
        transform_rot.transform.rotation.w = 0
```

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Nit
```suggestion
        # Create a transform combining a 100m z translation with
```

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Nit
```suggestion
        # So we create a transform with a 100m z translation
```

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Nit: You might want to set an authority such as `'unittest'` for this call, as seen elsewhere in this repository

https://github.com/ros2/geometry2/blob/bd125c5af32e2c63c70b7e0ba88e57094714b9b8/tf2_py/test/test_buffer_core.py#L112

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Nit: In case line 146 is changed in the future
```suggestion
            self.points -
            np.array([transform.transform.translation.x,
                transform.transform.translation.y,
                transform.transform.translation.z,])))
```

**Review by aprotyas — CHANGES_REQUESTED** [review]
Apparently this was left as a pending review and I just never submitted - thank you for the ping @Flova.

This PR looks good. I've left a bunch of nits inline, but feel free to ignore them if they don't make sense.

Consider what constitutes the public API of `tf2_sensor_msgs` though, i.e. my comments in the `__init__.py` file.

File: tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (Flova) [reply tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py:87]: Thanks ^^

File: tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py
Comment (Flova) [reply tf2_sensor_msgs/tf2_sensor_msgs/tf2_sensor_msgs.py:37]: Me neither, but it was the easiest fix for now. I guess it will only be removed once the used NumPy version includes this feature for all target systems, and in this case the above is imported. Overall, I would prefer some sort of repo which includes library feature ports like this for all components when it is needed due to old deps on some platforms.

**Comment by Flova** [comment]
@aprotyas everything should be resolved now

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Nit
```suggestion
        transform.transform.rotation.w = 0
```

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Nit
```suggestion
        transform.rotation.w = 0
```

File: tf2_sensor_msgs/test/test_tf2_sensor_msgs.py
Comment (aprotyas) [inline tf2_sensor_msgs/test/test_tf2_sensor_msgs.py]: Nit: single quotes preferred according to the style guide

```suggestion
        tf_buffer.set_transform_static(transform, 'unittest')
```

**Review by aprotyas — APPROVED** [review]
Thanks for iterating @Flova. I've left a couple of stray, minor nits that you can consider, but I'll approve regardless.

**Comment by aprotyas** [comment]
@Flova can you rebase your branch on the upstream HEAD? I can run CI for you once that's done.

**Comment by Flova** [comment]
Unfortunately trivial autocasing (e.g int -> float) for messages is removed in ros2 so one needs to write `0.0` instead of `0` when setting the values of the quaternion. I am currently on the go, but i will fix this later and do a rebase as well if I am back home.

**Comment by aprotyas** [comment]
Thanks for rebasing @Flova. Running CI now.

Repos file: https://gist.githubusercontent.com/aprotyas/25ed5c597ccf7fd55a64975b4a5c0abe/raw/0eb6fb2688926456e2944b39c07492898e9a8078/ros2.repos
Build args: `--packages-above-and-dependencies tf2_sensor_msgs`
Test args: `--packages-above tf2_sensor_msgs`
Job: https://ci.ros2.org/job/ci_launcher/10193/

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16588/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11170/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16974/)

**Comment by aprotyas** [comment]
@Flova looks like there are some flake8 linter warnings, can you address them please? For the set comprehension warning, going from `set(...)` to `{...}` should suffice.

**Comment by Flova** [comment]
@aprotyas I fixed the flake8 errors. Do you know why they only occur in this CI and not in the Jammy Build that is done automatically or my local installation?

**Comment by aprotyas** [comment]
@Flova it looks like the CI job installs these flake8 extensions which aren't present in the PR job. Same could apply for your local environment?

```
flake8==4.0.1
flake8-blind-except==0.1.1
flake8-builtins==1.5.3
flake8-class-newline==1.6.0
flake8-comprehensions==3.8.0
flake8-deprecated==1.3
flake8-docstrings==1.6.0
flake8-import-order==0.18.1
flake8-quotes==3.3.1
```

Re-running CI after 2cdf774.
Repos file: https://gist.githubusercontent.com/aprotyas/25ed5c597ccf7fd55a64975b4a5c0abe/raw/0eb6fb2688926456e2944b39c07492898e9a8078/ros2.repos
Build args: `--packages-above-and-dependencies tf2_sensor_msgs`
Test args: `--packages-above tf2_sensor_msgs`
Job: https://ci.ros2.org/job/ci_launcher/10194/

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16589/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11171/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16975/)

**Comment by aprotyas** [comment]
@clalancette feature PRs can now be merged into Rolling again right? If so, this PR is ready to go.

File: tf2_sensor_msgs/package.xml
Comment (clalancette) [inline tf2_sensor_msgs/package.xml]: These three new dependencies can all be `exec_depend`, I believe, since they are only ever used from Python.

(the geometry_msgs should be a `depend` since the C++ code is depending on it, we just missed it previously)

**Review by clalancette — CHANGES_REQUESTED** [review]
One more small thing to fix, then I think this is good.

**Comment by Flova** [comment]
The fix is applied. @clalancette

**Review by aprotyas — APPROVED** [review]
LGTM. Running CI again:

Repos file: https://gist.githubusercontent.com/aprotyas/25ed5c597ccf7fd55a64975b4a5c0abe/raw/0eb6fb2688926456e2944b39c07492898e9a8078/ros2.repos
Build args: `--packages-above-and-dependencies tf2_sensor_msgs`
Test args: `--packages-above tf2_sensor_msgs`
Job: https://ci.ros2.org/job/ci_launcher/10208/

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16603/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11187/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16993/)

**Comment by aprotyas** [comment]
Green CI + two approvals, so I'll merge this. Thanks for your contribution @Flova!

**Comment by Blast545** [comment]
:man_farmer: Looks like this PR introduced a build regression in the cyclonedds jobs of the buildfarm, see:

[Rci__nightly-cyclonedds_ubuntu_jammy_amd64#57](https://build.ros2.org/view/Rci/job/Rci__nightly-cyclonedds_ubuntu_jammy_amd64/57/)
[Rci__nightly-cyclonedds_ubuntu_jammy_amd64#58](https://build.ros2.org/view/Rci/job/Rci__nightly-cyclonedds_ubuntu_jammy_amd64/58/)


```
01:33:42.101 -- Found ament_lint_auto: 0.12.3 (/tmp/ws/install_isolated/ament_lint_auto/share/ament_lint_auto/cmake)
01:33:42.101 CMake Error at CMakeLists.txt:41 (find_package):
01:33:42.101   By not providing "Findament_cmake_pytest.cmake" in CMAKE_MODULE_PATH this
01:33:42.101   project has asked CMake to find a package configuration file provided by
01:33:42.101   "ament_cmake_pytest", but CMake did not find one.
01:33:42.101 
01:33:42.101   Could not find a package configuration file provided by
01:33:42.101   "ament_cmake_pytest" with any of the following names:
01:33:42.101 
01:33:42.101     ament_cmake_pytestConfig.cmake
01:33:42.101     ament_cmake_pytest-config.cmake
01:33:42.101 
01:33:42.101   Add the installation prefix of "ament_cmake_pytest" to CMAKE_PREFIX_PATH or
01:33:42.101   set "ament_cmake_pytest_DIR" to a directory containing one of the above
01:33:42.101   files.  If "ament_cmake_pytest" provides a separate development package or
01:33:42.101   SDK, be sure it has been installed.
01:33:42.101 
```

Can you take a look @Flova ?

**Comment by Blast545** [comment]
Mmm I think it only requires adding `ament_cmake_pytest` to the `package.xml` file, I'll test this out and open a PR, we can discuss there.

**Comment by aprotyas** [comment]
> Mmm I think it only requires adding `ament_cmake_pytest` to the `package.xml` file, I'll test this out and open a PR, we can discuss there.

@Blast545 https://github.com/ros2/geometry2/pull/520 should fix it.

**Comment by Blast545** [comment]
Just saw it, thanks! @aprotyas

**Pull request merged**
